internal/provider/provider.go

```diff
@@ -0,0 +1,818 @@
+				/*Validators: []validator.Bool{
+					boolvalidator.ConflictsWith(path.MatchRoot("password"),path.MatchRoot("private_key_path"), path.MatchRoot("private_key"),path.MatchRoot("private_key_passphrase"),path.MatchRoot("oauth_refresh_token"),path.MatchRoot("oauth_refresh_token")),
+				},*/
+			},
+			"private_key_path": schema.StringAttribute{
+				Description: "Path to a private key for using keypair authentication. Cannot be used with `browser_auth`, `oauth_access_token` or `password`. Can also be sourced from `SNOWFLAKE_PRIVATE_KEY_PATH` environment variable.",
+				Optional:    true,
+				Sensitive:   true,
+				Validators: []validator.String{
+					stringvalidator.ConflictsWith(path.MatchRoot("browser_auth"), path.MatchRoot("oauth_access_token"), path.MatchRoot("password")),
+				},
+				DeprecationMessage: "use the [file Function](https://developer.hashicorp.com/terraform/language/functions/file) instead",
+			},
+		},
+		Blocks: map[string]schema.Block{
+			"token_accessor": schema.ListNestedBlock{
+				Validators: []validator.List{
+					listvalidator.SizeAtMost(1),
+				},
+				NestedObject: schema.NestedBlockObject{
+					Attributes: map[string]schema.Attribute{
+						"token_endpoint": schema.StringAttribute{
+							Description: "The token endpoint for the OAuth provider e.g. https://{yourDomain}/oauth/token when using a refresh token to renew access token. Can also be sourced from the `SNOWFLAKE_TOKEN_ACCESSOR_TOKEN_ENDPOINT` environment variable.",
+							Required:    true,
+							Sensitive:   true,
+						},
+						"refresh_token": schema.StringAttribute{
+							Description: "The refresh token for the OAuth provider when using a refresh token to renew access token. Can also be sourced from the `SNOWFLAKE_TOKEN_ACCESSOR_REFRESH_TOKEN` environment variable.",
+							Required:    true,
+							Sensitive:   true,
+						},
+						"client_id": schema.StringAttribute{
+							Description: "The client ID for the OAuth provider when using a refresh token to renew access token. Can also be sourced from the `SNOWFLAKE_TOKEN_ACCESSOR_CLIENT_ID` environment variable.",
+							Required:    true,
+							Sensitive:   true,
+						},
+						"client_secret": schema.StringAttribute{
+							Description: "The client secret for the OAuth provider when using a refresh token to renew access token. Can also be sourced from the `SNOWFLAKE_TOKEN_ACCESSOR_CLIENT_SECRET` environment variable.",
+							Required:    true,
+							Sensitive:   true,
+						},
+						"redirect_uri": schema.StringAttribute{
+							Description: "The redirect URI for the OAuth provider when using a refresh token to renew access token. Can also be sourced from the `SNOWFLAKE_TOKEN_ACCESSOR_REDIRECT_URI` environment variable.",
+							Required:    true,
+							Sensitive:   true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func (p *SnowflakeProvider) Configure(ctx context.Context, req provider.ConfigureRequest, resp *provider.ConfigureResponse) {
```

Comment:
Parameter setting looks like it could be simplified by creating common functions for setting (at least) simple values. e.g.

```go
// instead of
config := &gosnowflake.Config{
	Application: "terraform-provider-snowflake",
}
account := os.Getenv("SNOWFLAKE_ACCOUNT")
if data.Account.ValueString() != "" {
	account = data.Account.ValueString()
}
if account != "" {
	config.Account = account
}
// we could something like this
config := &gosnowflake.Config{
	Application: "terraform-provider-snowflake",
	Account:     GetStringValue(data.Account, "SNOWFLAKE_ACCOUNT"),
	// Other params...
}

type StringValuer interface {
	ValueString() string
}

func GetStringValue(valuer StringValuer, envVariable string) string {
	if value := valuer.ValueString(); value != "" {
		return value
	}
	return os.Getenv(envVariable)
}
```
I know there can be different variations for the same type (e.g. with secondary env variable, with default value, etc.), but It's hard to clearly see what is going on and to check if every if statement is correct.

Reply:
...or a config builder

Reply:
a good idea, i also wasn't happy with the current design, but I was just trying to keep it as vanilla as possible. the current implementation matches how the tf provider docs suggest.